File: python-server/server/rag/chat/rag_engine.py
"""RAG 问答引擎（逐字移植自 src/application/chat/ragEngine.ts）。

流程：[可选]混合检索 → context 事件（pre-rerank 全量）→ Rerank（仅影响 LLM prompt）
→ 构建 prompt → LLM 流式回答；无 LLM 时产出 no-llm 事件降级。
"""

from __future__ import annotations

from collections.abc import Generator
from dataclasses import dataclass

from ..retrieve.hybrid_search import HybridSearchOptions
from ..types import SearchQuery, SearchResult
from .prompt_template import BuildOptions, PromptTemplate
from .types import (
    RagContextEvent,
    RagDoneEvent,
    RagErrorEvent,
    RagNoLlmEvent,
    RagTokenEvent,
)

_prompt_template = PromptTemplate()


@dataclass
class RagChatOptions:
    llm: object | None = None
    top_k: int = 5
    pre_search_results: list[SearchResult] | None = None
    entity_docs_content: str | None = None
    conversation_context: str | None = None
    is_follow_up: bool = False
    matched_keywords: list[str] | None = None
    rerank_top_k: int = 5


def _build_rag_prompt(query: str, search_results: list[SearchResult], options: dict):
    """检索结果按 score 降序，拼 `### 文档 i: 标题 (元信息)` + chunk.content。"""
    sorted_results = sorted(search_results, key=lambda r: -r.score)
    context_parts: list[str] = []

    for i, result in enumerate(sorted_results):
        chunk = result.chunk
        meta = chunk.metadata or {}

        header = f"### 文档 {i + 1}: {chunk.doc_title}"
        meta_parts = []
        if meta.get("client"):
            meta_parts.append(f"客户: {meta['client']}")
        if meta.get("project"):
            meta_parts.append(f"项目: {meta['project']}")
        if meta.get("docType"):
            meta_parts.append(f"类型: {meta['docType']}")
        if meta.get("date"):
            meta_parts.append(f"日期: {meta['date']}")
        if meta_parts:
            header += f" ({' | '.join(meta_parts)})"

        context_parts.append(f"{header}\n{chunk.content}")

    context = "\n\n---\n\n".join(context_parts)

    return _prompt_template.build(
        BuildOptions(
            context=context,
            query=query,
            conversation_context=options.get("conversation_context"),
            is_follow_up=bool(options.get("is_follow_up")),
            intent=options.get("intent"),
            entity_docs_content=options.get("entity_docs_content"),
        )
    )


def create_rag_chat_stream(deps: dict):
    default_llm = deps["llm"]
    reranker_factory = deps["reranker_factory"]
    hybrid_search = deps["hybrid_search"]

    def rag_chat_stream(
        query: str,
        options: RagChatOptions | None = None,
    ) -> Generator:
        options = options or RagChatOptions()
        top_k = options.top_k or 5
        llm = options.llm or default_llm

        # Step 1: 检索（有预检索结果/实体文档则跳过）
        if options.pre_search_results:
            search_results = list(options.pre_search_results)
            print(f"[RAG] 使用预检索结果: {len(search_results)} 个文档块")
        elif options.entity_docs_content:
            search_results = []
            print("[RAG] 已加载实体关联文档，跳过语义检索")
        else:
            try:
                search_results = hybrid_search(
                    query,
                    top_k,
                    20,
                    20,
                    HybridSearchOptions(matched_keywords=options.matched_keywords),
                )
                print(f"[RAG] 检索到 {len(search_results)} 个相关文档块")
            except Exception as err:
                print(f"[RAG] 检索失败: {err}")
                yield RagErrorEvent(content="文档检索失败，请检查知识库索引是否已初始化")
                return

        if not search_results and not options.entity_docs_content:
            yield RagErrorEvent(content="未找到相关文档，请尝试更换查询关键词")
            return

        # pre-rerank 全量结果回传前端展示
        yield RagContextEvent(results=search_results)

        # Step 1.5: Rerank（仅影响 LLM prompt）
        rerank_top_k = options.rerank_top_k or 5
        prompt_results = search_results
        if len(search_results) > rerank_top_k:
            try:
                reranked = reranker_factory().rerank(
                    SearchQuery(query=query), search_results, rerank_top_k
                )
                if reranked:
                    print(
                        f"[RAG] Rerank 重排序: {len(search_results)} → "
                        f"{len(reranked)} 个文档块（仅影响 LLM prompt）"
                    )
                    prompt_results = reranked
            except Exception as err:
                print(f"[RAG] Rerank 失败，使用原始结果: {err}")

        # Step 2: 构建 prompt
        system_prompt, user_prompt = _build_rag_prompt(
            query,
            prompt_results,
            {
                "entity_docs_content": options.entity_docs_content,
                "conversation_context": options.conversation_context,
                "is_follow_up": options.is_follow_up,
            },
        )

        # Step 3: 流式回答（无 LLM → no-llm 降级，不产出 done）
        if not getattr(llm, "available", False):
            yield RagNoLlmEvent(results=search_results)
            return

        try:
            stream = llm.stream(
                [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt},
                ]
            )
            for content in stream:
                if content:
                    yield RagTokenEvent(content=content)

            yield RagDoneEvent()
        except Exception as err:
            print(f"[RAG] LLM 调用失败: {err}")
            yield RagErrorEvent(
                content=f"LLM 调用失败: {str(err) or '未知错误'}"
            )

    return rag_chat_stream

File: src/application/chat/ragEngine.ts
// ============================================================
// RAG 问答引擎（Application 层 Use Case）
// 流程：用户问题 -> [可选]混合检索 -> Rerank -> 构建 prompt -> LLM 流式回答
// 支持多轮对话和上下文压缩
//
// 架构分层：
//   - LLM 调用走 LlmClient Port（OpenAI 实现在 infrastructure，由 Composition 注入）
//   - 重排走 Reranker Port（注入）
//   - 检索走注入的 hybridSearch Use Case
// ============================================================

import type { SearchResult, Reranker } from '@/domain/search/types';
import type { LlmClient } from '../ports';
import type { HybridSearchFn } from '../search/hybridSearch';
import { PromptTemplate } from './promptTemplate';

/** 提示词模板实例（复用） */
const promptTemplate = new PromptTemplate();

/** RAG 流事件 */
export interface RagChatEvent {
  type: 'context' | 'token' | 'done' | 'error' | 'no-llm';
  content?: string;
  results?: SearchResult[];
}

export interface RagChatOptions {
  /** 覆盖默认 LlmClient（前端自定义配置时由 route 创建传入） */
  llm?: LlmClient;
  topK?: number;
  /** 预检索结果，如果提供则跳过检索步骤 */
  preSearchResults?: SearchResult[];
  /** 实体关联的 Raw 文档全文内容 */
  entityDocsContent?: string;
  /** 对话历史上下文（用于多轮对话） */
  conversationContext?: string;
  /** 是否为追问 */
  isFollowUp?: boolean;
  /** 匹配到的实体关键字 */
  matchedKeywords?: string[];
}

export type RagChatStreamFn = (
  query: string,
  options?: RagChatOptions
) => AsyncGenerator<RagChatEvent>;

/** 构建 RAG Prompt（全量加载所有 chunk，不做截断） */
function buildRAGPrompt(
  query: string,
  searchResults: SearchResult[],
  options?: {
    entityDocsContent?: string;
    conversationContext?: string;
    isFollowUp?: boolean;
    intent?: string;
  }
): { systemPrompt: string; userPrompt: string } {
  // 构建文档上下文（混合检索结果 + 可选的数据库查询提示词）
  const sorted = [...searchResults].sort((a, b) => b.score - a.score);
  const contextParts: string[] = [];

  for (let i = 0; i < sorted.length; i++) {
    const result = sorted[i];
    const chunk = result.chunk;
    const meta = chunk.metadata;

    let header = `### 文档 ${i + 1}: ${chunk.docTitle}`;
    const metaParts: string[] = [];
    if (meta.client) metaParts.push(`客户: ${meta.client}`);
    if (meta.project) metaParts.push(`项目: ${meta.project}`);
    if (meta.docType) metaParts.push(`类型: ${meta.docType}`);
    if (meta.date) metaParts.push(`日期: ${meta.date}`);
    if (metaParts.length > 0) {
      header += ` (${metaParts.join(' | ')})`;
    }

    contextParts.push(`${header}\n${chunk.content}`);
  }

  const context = contextParts.join('\n\n---\n\n');

  return promptTemplate.build({
    context,
    query,
    conversationContext: options?.conversationContext,
    isFollowUp: options?.isFollowUp,
    intent: options?.intent,
    entityDocsContent: options?.entityDocsContent,
  });
}

/**
 * 创建流式 RAG 回答 Use Case
 *
 * @param deps - llm（LlmClient Port，默认实例）+ rerankerFactory（按需创建 Reranker）+ hybridSearch
 */
export function createRagChatStream(deps: {
  llm: LlmClient;
  rerankerFactory: () => Reranker;
  hybridSearch: HybridSearchFn;
}): RagChatStreamFn {
  const { llm: defaultLlm, rerankerFactory, hybridSearch } = deps;

  return async function* ragChatStream(
    query: string,
    options?: RagChatOptions
  ): AsyncGenerator<RagChatEvent> {
    const topK = options?.topK || 5;
    const llm = options?.llm ?? defaultLlm;

    // Step 1: 检索（如果已有预检索结果则跳过；如果有实体文档内容也跳过）
    let searchResults: SearchResult[];
    if (options?.preSearchResults && options.preSearchResults.length > 0) {
      searchResults = options.preSearchResults;
      console.log(`[RAG] 使用预检索结果: ${searchResults.length} 个文档块`);
    } else if (options?.entityDocsContent) {
      // 有实体关联文档内容时，跳过语义检索，直接用空结果（prompt 中会以实体文档为主）
      searchResults = [];
      console.log('[RAG] 已加载实体关联文档，跳过语义检索');
    } else {
      try {
        searchResults = await hybridSearch(query, topK, 20, 20, {
          matchedKeywords: options?.matchedKeywords,
        });
        console.log(`[RAG] 检索到 ${searchResults.length} 个相关文档块`);
      } catch (err) {
        console.error('[RAG] 检索失败:', err);
        yield { type: 'error', content: '文档检索失败，请检查知识库索引是否已初始化' };
        return;
      }
    }

    // 检查检索结果（如果有实体文档内容，允许跳过语义检索结果）
    if (searchResults.length === 0 && !options?.entityDocsContent) {
      yield { type: 'error', content: '未找到相关文档，请尝试更换查询关键词' };
      return;
    }

    // 返回检索上下文给前端展示（Rerank 之前，保留完整数量）
    yield { type: 'context', results: searchResults };

    // Step 1.5: Rerank 重排序（语义相关性精排，仅用于 LLM prompt，不影响前端展示）
    let promptResults = searchResults;
    if (searchResults.length > 5) {
      try {
        const rerankedResults = await rerankerFactory().rerank({ query }, searchResults, 5);
        if (rerankedResults.length > 0) {
          console.log(`[RAG] Rerank 重排序: ${searchResults.length} → ${rerankedResults.length} 个文档块（仅影响 LLM prompt）`);
          promptResults = rerankedResults;
        }
      } catch (err) {
        console.warn('[RAG] Rerank 失败，使用原始结果:', err);
      }
    }

    // Step 2: 构建 prompt（传入对话历史上下文和追问标记，使用精排后的结果）
    const { systemPrompt, userPrompt } = buildRAGPrompt(query, promptResults, {
      entityDocsContent: options?.entityDocsContent,
      conversationContext: options?.conversationContext,
      isFollowUp: options?.isFollowUp,
    });

    // Step 3: 调用 LLM 流式输出（无 LLM 时降级为 no-llm 事件，由前端处理展示）
    if (!llm.available) {
      yield { type: 'no-llm', results: searchResults };
      return;
    }

    try {
      const stream = llm.stream({
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userPrompt },
        ],
      });

      for await (const content of stream) {
        if (content) {
          yield { type: 'token', content };
        }
      }

      yield { type: 'done', results: searchResults };
    } catch (err: unknown) {
      const error = err instanceof Error ? err : new Error(String(err));
      console.error('[RAG] LLM 调用失败:', error);
      yield { type: 'error', content: `LLM 调用失败: ${error.message || '未知错误'}` };
    }
  };
}
